juju/juju
# feat: juju switch to previous controller or model

It implements a new behavior to `juju switch`, similar to well known shell command `cd -`

It also introduce a flag-based way to switch, through `-c` and `-m` flags (as done in other command):
* `juju switch -m a-model` equivalent to `juju switch :a-model`
* `juju switch -c a-controller` equivalent to `juju switch a-controller:`

To implements the behavior similar with `cd -`, this PR introduce the placeholder `-` which mean "previous". So:

* `juju switch -` # switch to previous state controller:model (current model of previous controller if the controller has changed, previous model of current controller if model has changed)
* `juju switch -m -` # switch to previous model, current controller
* `juju switch -c -` # switch to previous controller

I also added two related but independent commit:

* A commit to remove a duplicate TU for the switch command
* A commit which remove several golang warning from the files I have modified.

## Checklist




## QA steps

Unit test: `TEST_PACKAGES="./cmd/juju/commands/... -count=1 -race -check.vv" make run-go-tests`

However jujucli state is stored in a file which use handled by [jujuclient/file.go] which has no UT (and the store implemented there is replace by a in-memory store and thus uncovered). It requires few manual testing.

```sh
# providing we have 
# ctrl1 with m1 & m2
# ctrl2 with m1 & m2

# Arrange
juju switch ctrl1:m1 # set previous model for ctrl1
juju switch m2  # set current model for ctrl1
juju switch ctrl2:m1  # set previous model for ctrl2
juju switch m2  # set current model for ctrl2

# Assert
juju switch -c - # ctrl2:m2 -> ctrl1:m2 (switch controller)
juju switch -m - # ctrl1:m2 -> ctrl1:m1 (switch model)
juju switch - # ctrl1:m1 -> ctrl1:m2 (switch model)
juju switch ctrl2:m1 # ctrl1:m2 -> ctrl2:m1 (switch controller & model)
juju switch - # ctrl2:m1 -> ctrl1:m2 (switch controller)
juju add-model m3 # ctrl1:m2 -> ctrl1:m3 (switch model)
juju switch - # ctrl1:m3 -> ctrl1:m2 (switch model)
```

## Documentation changes

[juju switch](https://juju.is/docs/juju/juju-switch) command documentation need to be updated 

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/1988426

**Jira card:** [JUJU-6168](https://warthogs.atlassian.net/browse/JUJU-6168)

[JUJU-6168]: https://warthogs.atlassian.net/browse/JUJU-6168?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
